apps/prairielearn/src/ee/pages/lti13CourseNavigation/lti13CourseNavigation.ts

```diff
@@ -1,10 +1,189 @@
+} from './lti13CourseNavigation.html';
+import { Lti13Claim } from '../../lib/lti13';
 
+const sql = loadSqlEquiv(__filename);
 const router = Router({ mergeParams: true });
 
-// Placeholder to be enlarged later
-router.all('/', (_req, res) => {
-  res.redirect('/pl');
-});
+router.get(
+  '/',
+  asyncHandler(async (req, res) => {
+    if ('done' in req.query) {
+      res.send(
+        Lti13CourseNavigationDone({
+          resLocals: res.locals,
+          lti13_instance_id: req.params.lti13_instance_id,
+        }),
+      );
+      return;
+    }
+
+    const LTI = new Lti13Claim(req);
+    const courseName = `${LTI.context?.label}: ${LTI.context?.title}`;
+    const role_instructor = LTI.isRoleInstructor();
+
+    // Get lti13_course_instance info, if present
+    const lci = await queryOptionalRow(
+      sql.get_course_instance,
+      {
+        lti13_instance_id: req.params.lti13_instance_id,
+        deployment_id: LTI.deployment_id,
+        context_id: LTI.context?.id,
+      },
+      Lti13CourseInstanceSchema,
+    );
+
+    if (lci) {
+      if (role_instructor) {
+        // Update lti13_course_instance on instructor login, helpful as LMS updates or we add features
+        await queryAsync(sql.upsert_lci, {
+          lti13_instance_id: req.params.lti13_instance_id,
+          course_instance_id: lci.course_instance_id,
+          deployment_id: LTI.deployment_id,
+          context_id: LTI.context?.id,
+          context_label: LTI.context?.label,
+          context_title: LTI.context?.title,
+        });
+
+        // TODO: Set course/instance staff permissions for LMS course staff here?
+      }
+
+      // LTI claims are not used after this page so remove them from the session
+      LTI.remove();
+
+      // Redirect to linked course instance
+      res.redirect(
+        `/pl/course_instance/${lci.course_instance_id}/${role_instructor ? 'instructor/' : ''}`,
+      );
+      return;
+    }
+
+    if (!role_instructor) {
+      // Students get a "come back later" message
+
+      /*
+       * TODO: to not break things for current testing courses,
+       *       fall back to the previous PR behavior of simply
+       *       redirecting to /pl
+       */
+      res.redirect('/pl/');
+      return;
+
+      /*
+      res.send(
+        Lti13CourseNavigationNotReady({
+          resLocals: res.locals,
+          courseName,
+        }),
+      );
+      return;
+      */
+    }
+
+    // Instructor so lookup their existing information in PL
+    const courses = await selectCoursesWithEditAccess({
+      user_id: res.locals.authn_user.user_id,
+      is_administrator: res.locals.authn_is_administrator,
+    });
+
+    let course_instances: CourseInstance[] = [];
+
+    for (const course of courses) {
+      // Only course owners can link
+      if (!course.permissions_course.has_course_permission_own) {
+        continue;
+      }
+
+      const loopCI = await selectCourseInstancesWithStaffAccess({
```

Comment:
Let's a get a better variable name than `loopCI`. Just `instances` would be better.